# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [ ]:
# def addTwo(x):
#     return x + 2

# print(addTwo(4))

# # Lambda Function Syntax
# # Example of calling a lambda function without a variable
# print((lambda x: x + 2)(4))


In [1]:
def add_ten(x):
    return x +10
print(add_ten(4))

add_something = lambda x :x+6
print(add_some_thing(7))

print(lambda y : y*4))



14


NameError: name 'add_some_thing' is not defined

In [5]:
x = lambda a: a*10
print(x(10))

y= lambda name: print(name)
y('Boba Fett')

z = lambda num : int(num) + 80

print(z('40'))




100
Boba Fett
120


#### Saving to a Variable

In [ ]:
f_test = lambda x: x + 2

f_test(4)

#### Multiple Inputs

In [ ]:
# # Multiple Inputs with no variable
# print((lambda x,y,z: x * y * z)(3,5,8))

# # Multiple Inputs with a variable attached
# x_test = lambda x,y,z: x * y * z
# print(x_test(3,5,8))

In [10]:
print((lambda x, y, z: x*y*z)(3, 5, 8))

x_test = lambda x,y,z: x*y+z

print(x_test(5,8,23))

string_test = lambda verb, adjective, noun: f"It is a {adjective}{noun} and I {verb} to the store."

print(string_test('roll','side walk','tasty'))


120
63
It is a side walktasty and I roll to the store.


#### Passing a Lambda into a Function

In [6]:
# def multiply(f,num):
#     """
#         f expects a lambda function
#         num expects a number
#     """
#     return f(num)
# multiply(lambda x: x * x, 4) # lambda 4: 4 * 4

16

In [13]:
def multiply(f,num):
    """
#         f expects a lambda function
#         num expects a number
#     """
    return f(num)
multiply(lambda x:x*4,6)

def multiply2(num):
    f = lambda x:x*4
    return f(num)
print(multiply2(7))


28


#### Returning a Lambda from a Function

In [ ]:
# # Regular defined function
# def multiply_test(num):
#     return num * 4

# # Function within a function
# def returnFunc():
#     test = 4
#     def multiply(num):
#         return num * 2
#     return multiply

# f_return = returnFunc()
# print(returnFunc())
# print(f_return(4))

# # Lambda function returned from a regular function
# def returnLam(b,c):
#     return lambda x,a: x + a + b + c
# r_lamb = returnLam(4,5)
# print(r_lamb(5,5))

In [20]:
def multiply_test(num):
     return num * 4
    
def returnFunc():
    test = 4
    def multiply(num):
        return num * 2
    return multiply
func_return = returnFunc()
#print(func_return)
print(func_return(8))

def add_num(num):
    return num + 3

# # Lambda function returned from a regular function
def return_lamb(b,c):
    return lambda x, a: x + a + b + c
    
r_lamb = return_lamb(4,5)
print(return_lamb(4,5)(7,8))

print(r_lamb(9,10))


16
24
28


#### If Statements within Lambdas

In [ ]:
# # Lambda x : True if (condition) else False
# f_condition = lambda num : num * 2 if num > 10 else num + 2

# print(f_condition(8))

# print(f_condition(12))

# print(f_condition(10))

In [25]:
# Lambda x : True if (condition) else False
# Electric Company
# Create a function that given a list which represents street lights given as a parameter(l_street), determine
#  if an outage has occurred. A street with a total number of "F" greater than or equal to 2 returns "Outage", anything below returns "Power"
# Example Input: [ 'T', 'F', 'F', 'F' ]
# Example Output: "Outage"

# is_there_power = lambda arr: "outage" is arr.count('F') >= 2 else "Power"
# print(is_there_power([ 'T', 'F', 'F', 'F' ]))




f_condition = lambda num: num*2 if num >10 else num + 2
print(condition(8))

print(f_condition(10))

print(f_codition(12))
                                                  


NameError: name 'condition' is not defined

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [3]:
if they are less:

# f = lambda cube: cube ** 3

# print(f(2))
# print(f(3))
# print(f(4))


func_cond = landa num: num*2 if num  >= 15 
    f = (func_cond)
print(f(2))
print(f(3))
print(f(4))  


    


SyntaxError: invalid syntax (207412392.py, line 1)

In [ ]:
anonymous_function = lambda num: num**3 if num < 15 else num**2

x = anonymous_function

y = lambda a, b: a*b

print(x(5))

print(x(20))

print(x(y(2,5)))

In [29]:
anonymous_function = lambda num: num**3 if num < 15 else num**2

x = anonymous_function

y = lambda a, b: a*b

print(x(5))

print(x(20))

print(x(y(2,5)))

125
400
1000


In [32]:
f = lambda x: x**3 if x <15 else x**2
print(f(5))

125


In [33]:
f = lambda x: x**3 if x < 15 else x**2
print(f(10))
print(f(20))

1000
400


In [34]:
f = lambda num: num ** 3 if num < 15 else num ** 2

print(f(20))

400


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [ ]:
# # map(func,iterable(list,dict,tuple,set))
# # Normally the usage of map happens with a pre-defined function -- but we can use lambdas as well
# def squared(num,num2):
#     if num < 10 and num2 < 10:
#         return num ** 2, num2 ** 2
#     else:
#         return num,num2
    
# numbers = [4,11,20,3,15,20]
# more_nums = [4,10,3,2,6]

# squared_nums_map = list(map(squared,numbers,more_nums))
# print(squared_nums_map)






In [37]:
def square_nums(num):
    return num **2
arr =[2,3,4,5,6,7] 
for x in arr:
    squared_list.append(square_nums(x))
print(squared_list)

defsquared(num):
    if num <10:
        return num **2
    else:
        return num
squared_nums_map = = list(map((squared, arr)))
print(squared_nums_map)

NameError: name 'squared_list' is not defined

In [39]:
arr = [2, 30, 4, 15, 6, 17]
arr2 = [7, 18, 4, 16, 87]
def squared2(num, num2):
    if num < 10 and num2 < 10:
        return num **2, num2 **2
    else:
        return num, num2
squared_num_map = list(map(squared2, arr, arr2))
print(squared_num_map)
        


[(4, 49), (30, 18), (16, 16), (15, 16), (6, 87)]


#### Using Lambda's with Map

In [ ]:
# map(lambda x: x + 2, list)
# using lambda in map happens in line (or in one line) usually
# list(map(lambda x : x * 2,nums))
squared_nums_lamb = list(map(lambda x,y : (x ** 2,y ** 2) if x < 10 and y < 10 else (x,y),numbers,more_nums))
print(squared_nums_lamb)





## 

In [41]:
arr = [2, 30, 4, 15, 6, 17]
arr2 = [7, 18, 4, 16, 87]

squared_nums_lamb = list(map(lambda x,y : (x ** 2,y ** 2) if x < 10 and y < 10 else (x,y),arr, arr2))
print(squared_nums_lamb)

[(4, 49), (30, 18), (16, 16), (15, 16), (6, 87)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [ ]:
double_num = list(map(lambda x,y: (((x * 2) - 1),((y * 2) - 1)) if x > 0 and y > 0 else (x,y), numbers,more_nums))
print(double_num)

In [44]:
arr = [2, 30, 4, 15, 6, 17]
double_num = list(map(lambda x,y: (((x * 2) - 1),((y * 2) - 1)) if x > 0 and y > 0 else (x,y), arr,num))
print(double_num)


NameError: name 'num' is not defined

In [ ]:
double_num = list(map(lambda x,y: (((x * 2) - 1),((y * 2) - 1)) if x > 0 and y > 0 else (x,y), numbers,more_nums))
print(double_num)

In [45]:
arr = [2, 30, 4, 15, 6, 17]

new_map = list(map(lambda x: x*2-1, arr))
print(new_map)

[3, 59, 7, 29, 11, 33]


In [46]:
arr = [2, 30, 4, 15, 6, 17]
double = list(map(lambda x: x * 2 - 1, arr))
print(double)

[3, 59, 7, 29, 11, 33]


In [47]:
y = lambda x: x*2 - 1
math_double_substractOne = list(map(lambda x: x*2 - 1, arr))
print(math_double_substractOne)

[3, 59, 7, 29, 11, 33]


In [48]:
new_map = list(map(lambda x: x*2-1, arr))
print(new_map)

[3, 59, 7, 29, 11, 33]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [ ]:
# names = ['Bob','Andy','Max','Evan','Angelica']

# def a_names(name):
#     if name[0].lower() == 'a':
#         return True
#     else:
#         return False
# new_names_list = list(filter(a_names, names))
# print(new_names_list)

# print(a_names)



In [ ]:
names = ['Bob','Andy','Max','Evan','Angelica']
def a_names(name):
    
    if name[0].lower() == 'a':
        return True
    else:
        return False
new_names_list = list(filter(a_names, names))
print(new_names_list)




#### Using Lambda's with Filter()

In [51]:
names = ['Bob','Andy','Max','Evan','Angelica']
new_names_lamb = list(filter(lambda name: True if name[0].lower() == 'a' else False, names))
print(new_names_lamb)
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
even_numbers = list(filter(lambda num: True if num % 2 == 0 else False, numbers))
print(even_numbers)


['Andy', 'Angelica']
[2, 4, 6, 8, 10]


In [59]:
from statistics import mean

# nums = [2, 7, 4.2, 1.6, 9, 4.4, 5.9]

# mean = list(mean(lambda nums: True if num > mean == 0 else False, nums))

# print(mean(nums))

In [61]:
from statistics import mean

nums = [2, 7, 4.2, 1.6, 9, 4.4, 5.9]

print(mean(nums))
print(list(filter(lambda num: num if num < mean(nums) else False, nums)))

4.871428571428572
[2, 4.2, 1.6, 4.4]


In [60]:
below_mean = list(filter(lambda a: True if a <= mean(nums) else False, nums))
print(below_mean)

[2, 4.2, 1.6, 4.4]


In [62]:
print(list(filter(
(lambda x: True if x <= mean(nums) else False),nums
)))

[2, 4.2, 1.6, 4.4]


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [63]:
from statistics import mean

nums = [2,7,4.2,1.6,9,4.4,4.9]

print(mean(nums))

avg = mean(nums)
print(avg)

def mean_test(name):
    if name < avg:
        return True
    else:
        return False
new_calc = list(filter(mean_test,nums))
print(new_calc)

# Lambda function solution
below_avg = list(filter(lambda x: True if x <= mean(nums) else False, nums))
print(below_avg)

4.728571428571429
4.728571428571429
[2, 4.2, 1.6, 4.4]
[2, 4.2, 1.6, 4.4]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [64]:
from functools import reduce

# reduce(function,iterable)

list_1 = [2,4,6,8,10]

def addNums(num1,num2):
    return num1 + num2

result_add = reduce(addNums,list_1)

print(result_add)

# Subtract a list of numbers
def subtractNums(num1,num2):
    return num1 - num2

result_sub = reduce(subtractNums, list_1)
print(result_sub)


30
-26


#### Using Lambda's with Reduce()

In [ ]:
result_lamb = reduce(lambda x,y: x + y, list_1)
print(result_lamb)

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [68]:
# my_list = [1,2,3,4]

# from functools import reduce

# result_lamb = reduce(lambda x,y: x * y, my_list)
# print(result_lamb)

list_1 = [2,4,6,8,10]

multiply_lamb = reduce(lambda x,y: x*y), list_1


TypeError: reduce expected at least 2 arguments, got 1

In [69]:

list_1 = [2,4,6,8,10]
result_lamb = reduce(lambda x, y: x * y, list1)
print(result_lamb)

NameError: name 'list1' is not defined

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [ ]:
def addNums(num):
    # Set base case for recusive function
    if num <= 1:
        print("addNums(1) = 1")
        return num
    else:
        print(f"addNums({num}) = {num} + addNums({num - 1})")
        return num + addNums(num - 1)
    
addNums(5)

#### Writing a Factorial Function

In [ ]:
# 5! = 5 * 4 * 3 * 2 * 1
def factorial(num):
    if num <= 1:
        return 1
    else:
        return num * factorial(num - 1)
    
factorial(5)


#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [ ]:
# return num - subNums(num - 1)
# subtract(10)
def subNums(num):
    if num <= 1:
        print("subNums(1) = 1")
        return num
    else:
        print(f"subNums({num}) = {num} - subNums({num - 1})")
        return num - subNums(num - 1)
    
subNums(5)

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [5]:
def my_range(stop, start, step = 2):
    while start < stop:
        yield start # yield keyword denotes a generator
        start += step
for i in my_range(20, start = 2):
    my_generator_value = i
    print(my_generator_value)

#my_range(20,start = 2)




2
4
6
8
10
12
14
16
18


#### next() keyword
<p>Returns next element from the list, if not present prints the default value. If default value is not present, raises the StopIteration error.</p>

In [6]:
def my_range(stop, start, step = 2):
    while start < stop:
        yield start # yield keyword denotes a generator
        start += step

#### Creating a Generator Object
<p>The generator object will store an iterable on which we can call the next() method.</p>

In [ ]:
generator_object = my_range(20, start = 2)

In [11]:
next(generator_object)

6

<p>Use try and except to print graceful message instead of the StopIteration error:

In [8]:
generator_object = my_range(20, start = 2)

In [19]:
try:
    print(next(generator_object))
except:
    print("You're out of range")

You're out of range


In [18]:
next(generator_object)

StopIteration: 

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [23]:
# for i in numbers_squared(10)

def num_sq(stop,start,step = 1):
    while start >= stop:
        yield start ** 2
        start -= step
        
for i in num_sq(0,start = 3):
    print(i)

9
4
1
0


In [26]:
def squared(stop, start, step = 1):
    while start >= stop:
        yield start ** 2
        start -= step
gen = squared(0, start = 25)

try: print(next(gen))
except: print("number has reached 0")

625


In [24]:
def my_range(stop, start, step):
    while start >= 0:
        yield start**2
        start -= step
for i in my_range(0,10,1):
    my_generator_value = i
    print(my_generator_value)

100
81
64
49
36
25
16
9
4
1
0


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [4]:
cities = [" ", "Argentina", " ", "San Diego", "", "  ", "", "Boston", "New York"]


cities_list = []


for city in cities:

   
    if city.strip() != "":

       
        cities_list.append(city.strip())


print(cities_list)



['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [3]:
authors = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

authors.sort(key=lambda name: name.split()[-1])

print(authors)

['Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'Victor aNisimov', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [19]:
#temperatures = [('Nashua', 32), ('Boston', 12), ('Los Angelos', 44), ('Miami', 29)]
# F = (9/5)*C + 32

c_temps = [('Nashua', 32), ('Boston', 12), ('Los Angelos', 44), ('Miami', 29)]

f_temps = list(map(lambda x: (x[0], (x[1] * 9/5) + 32), c_temps))

print(f_temps)


[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Exercise #4
¶
Create a generator function that individually returns each movie genre back from the list. Use a try and except to print an error message when the StopIteration error is hit</p>




In [ ]:
# genres = ["adventure", "drama", "horror", "comedy", "action", "romance"]



In [1]:
def genre_list():
    
    genres = ["adventure", "drama", "horror", "comedy", "action", "romance"]
    
    for genre in genres:
        
        yield genre
try:
    
    genre_generator = genre_list()
    
    while True:
        
        genre = next(genre_generator)
        
        print(genre)
        
except StopIteration:
    
    print("End of List")

adventure
drama
horror
comedy
action
romance
End of List
